packages/core/utils/lib/sanitize/sanitizers.js

```diff
@@ -16,7 +14,12 @@ const sanitizePrivates = curry((schema, entity) => {
 });
 
 const defaultSanitizeOutput = curry((schema, entity) => {
-  return pipeAsync(sanitizePrivates(schema), sanitizePasswords(schema))(entity);
+  const defaultSanitize = (schema, entity) => {
```

Comment:
I understand the idea but I don't think this will work.

Both visitors (removePasswords and removePrivate) expects a context and visitor utils as parameters, not the schema and the entity. :slightly_smiling_face:

Reply:
However you can create a composed visitor that'll combine both logic

```js
const removePrivateAndPasswords = (ctx, utlis) => {
  removePassword(ctx, utils);
  removePrivates(ctx, utils);
);
```

And then traverse the entity with that

Reply:
However you'll have an issue if you've a private password field, it'll try to remove it once, then try to remove it again even if it doesn't exists anymore

Reply:
Introducing a composeVisitor util might be a more viable solution

Reply:
a compose visitor pattern is one that one where a visitor calls the next one?'
For instance if removePassword removes the password, it would not call removePrivates.

Reply:
added your change. The user & permissions User seems to have the 'password' field set as private too. And it seems to be working.

Seems if you do `delete obj[key]` and the key does not exist in the object, it does not throw an error :)

Reply:
I am super interested on learning the visistor pattern and more examples about this. I will try to read a bit about the topic.

Reply:
> a compose visitor pattern is one that one where a visitor calls the next one?'

I don't know if "compose visitor pattern" really means something in the books, but yeah that would be the idea. I think it would make sense if we want/know each sanitize/transform operation to be independent from the other ones :thinking: 

> For instance if removePassword removes the password, it would not call removePrivates.

Indeed, but then it asks the question of: How do I know if I should run the chained visitor? If the value is updated by the first visitor I might still want to update or remove it in the next visitor. However if the key/value is removed, should I still perform operations in the second visitor? 

> Seems if you do delete obj[key] and the key does not exist in the object, it does not throw an error :)

The only issue with that is that we're still evaluating the original value, if the first visitor update the value, you won't have it up-to-date in the second visitor

RefactoringGuru's explanation of the visitor pattern is quite good imo: https://refactoring.guru/design-patterns/visitor

Reply:
I was reading the visitor pattern and wondering, do we really implement it?

Tell me if this is too picky, this is me trying to understand the pattern!

An example in the article of the pattern is:
```js
class Circle implements Shape is
    // ...
    method accept(v: Visitor) is
        v.visitCircle(this)

class Rectangle implements Shape is
    // ...
    method accept(v: Visitor) is
        v.visitRectangle(this)

class CompoundShape implements Shape is
    // ...
    method accept(v: Visitor) is
        v.visitCompoundShape(this)
```

What I see is a list of visitor methods that implements different actions for each shape. But we don't really do that no? The concrete element (following the article wording) I guess would be the traverse entity method, and inside that method, we are always calling the same method, not one depending on the implementation of the method.

What I feel we are using is the repository pattern, where the logic of the visitor is completely decoupled from the concrete element.

```js
class CompoundShape implements Shape is
    // ...
    method accept(repository: Repository) is
        repository(someData)
```

And theoretically, the repository could chain multiple key removals if necessary. The same as the compound visitor. @Convly